pkg/operator/certrotationcontroller/satokensigner_controller.go

```diff
@@ -270,6 +276,30 @@ func (c *SATokenSignerController) syncWorker() error {
 	return nil
 }
 
+func isValidKeyPair(pubKey, privKey []byte) (bool, error) {
+	privKeyDer, privBytesLeft := pem.Decode(privKey)
+	if privKeyDer == nil || len(privBytesLeft) != 0 {
+		return false, nil
+	}
+
+	pubKeyDer, pubBytesLeft := pem.Decode(pubKey)
+	if pubKeyDer == nil || len(pubBytesLeft) != 0 {
+		return false, nil
+	}
+
+	rsaPrivateKey, err := x509.ParsePKCS1PrivateKey(privKeyDer.Bytes)
+	if err != nil {
+		return false, err
+	}
+
+	rsaPublicKey, err := x509.MarshalPKIXPublicKey(rsaPrivateKey.Public())
+	if err != nil {
+		return false, err
+	}
+
+	return bytes.Compare(pubKeyDer.Bytes, rsaPublicKey) == 0, nil
```

Comment:
What is this check for?

Also, the check thus far has confirmed that the keys can be decoded and parsed. What guarantee is there that a signature generated with the private key can be validated with the public?

Reply:
function tries to extract public key from the given private key and see if that key matches the public key passes as argument. `bytes.Compare` is trying to match the public keys.If the public keys match, then it means that tokens signed with private key can be validated using public key

Reply:
TIL that RSA private keys also contain contain the public key in most implementations (as per https://tools.ietf.org/html/rfc3447#appendix-A.1.2). 

That suggests that there is no point to keeping the public key as a separate entry in the signing secret since it duplicates data that already appears in the private key.  `ParsePKCS1PrivateKey` validates that the public and private exponents are related, and that check has to be duplicated for the external public key (as you're doing here). I'm not suggesting changing that in this PR, but it does represent a future cleanup opportunity. I'm assuming the signing secret is only composed this way because whoever wrote the code originally confused rsa keypair management with CA management (since the private key does not embed the CA certificate).

Any reason you chose to marshal the private key's public key for comparison vs parsing the public key with `x509.ParsePKIXPublicKey`?

Reply:
> Any reason you chose to marshal the private key's public key for comparison vs parsing the public key with x509.ParsePKIXPublicKey?

I want bytes for comparison and marshal of the private key was giving me that. x509.ParsePKIXPublicKey would have given me interface and I need to assert it to convert it again to key format from interface format. That was my thinking at that time. It could as well could be done if I choose to use interfaces. Which I am using now in the recent commit.

Reply:
> That suggests that there is no point to keeping the public key as a separate entry in the signing secret since it duplicates data that already appears in the private key. ParsePKCS1PrivateKey validates that the public and private exponents are related, and that check has to be duplicated for the external public key (as you're doing here). I'm not suggesting changing that in this PR, but it does represent a future cleanup opportunity. I'm assuming the signing secret is only composed this way because whoever wrote the code originally confused rsa keypair management with CA management (since the private key does not embed the CA certificate).

I could not follow what you were saying. If I understand correctly, `kube-controller-manager` only cares about private key and it is responsible for signing keys.
https://github.com/kubernetes/kubernetes/blob/v1.17.0/cmd/kube-controller-manager/app/controllermanager.go#L556-L559
And if the function is supposed to validate key pair, so as far as I could think, there is no way of doing this without the public key passed as argument to this function. Please let me know if I am missing something.

Reply:
The private key entry in the secret embeds the public key. Therefore if the private key entry can be parsed without error but the public key entry in the secret cannot be parsed or does not match the public key embedded in the private key, it should not be necessary to create a new keypair. Just write the public key embedded in the private key to the public key entry in the secret. Feel free to contact me on slack if this is not clear.

Reply:
Ah! I get what you are saying now. Thanks!